datafusion/optimizer/src/rewrite_cycle.rs

```diff
@@ -0,0 +1,412 @@
+    /// let (evaluated_expr, info) = RewriteCycle::new()
+    ///     .with_max_cycles(4)
+    ///     .each_cycle(expr, |cycle_state| {
+    ///         cycle_state
+    ///             .rewrite(&mut addition_rewriter)?
+    ///             .rewrite(&mut multiplication_rewriter)
+    ///     })
+    ///     .unwrap();
+    /// assert_eq!(evaluated_expr, lit(74));
+    /// assert_eq!(info.completed_cycles(), 3);
+    /// assert_eq!(info.total_iterations(), 7);
+    /// ```
+    pub fn each_cycle<
+        Node: TreeNode,
+        F: FnMut(
+            RewriteCycleState<Node>,
+        ) -> RewriteCycleControlFlow<RewriteCycleState<Node>>,
+    >(
+        &self,
+        node: Node,
+        mut f: F,
+    ) -> Result<(Node, RewriteCycleInfo)> {
+        let mut state = RewriteCycleState::new(node);
+        if self.max_cycles == 0 {
+            return state.finish();
+        }
+        // run first cycle then record number of rewriters
+        state = match f(state) {
+            ControlFlow::Break(result) => return result?.finish(),
+            ControlFlow::Continue(node) => node,
+        };
+        state.record_cycle_length();
+        if state.is_done() {
+            return state.finish();
+        }
+        // run remaining cycles
+        match (1..self.max_cycles).try_fold(state, |state, _| f(state)) {
+            ControlFlow::Break(result) => result?.finish(),
+            ControlFlow::Continue(state) => state.finish(),
+        }
+    }
+}
+
+/// Iteration state of a rewrite cycle. See [RewriteCycle::each_cycle] for usage examples and information.
+#[derive(Debug)]
+pub struct RewriteCycleState<Node: TreeNode> {
+    node: Node,
+    consecutive_unchanged_count: usize,
+    rewrite_count: usize,
+    cycle_length: Option<usize>,
+}
+
+impl<Node: TreeNode> RewriteCycleState<Node> {
+    fn new(node: Node) -> Self {
+        Self {
+            node,
+            cycle_length: None,
+            consecutive_unchanged_count: 0,
+            rewrite_count: 0,
+        }
+    }
+
+    /// Records the rewrite cycle length based on the current iteration count
+    ///
+    /// When the total number of writers is not known upfront - such as when using
+    /// [RewriteCycle::each_cycle] we need to keep count of the number of [Self::rewrite]
+    /// calls and then record the number at the end of the first cycle.
+    fn record_cycle_length(&mut self) {
+        self.cycle_length = Some(self.rewrite_count);
+    }
+
+    /// Returns true when the loop has reached the maximum cycle length or when we've received
+    /// consecutive unchanged tree nodes equal to the total number of rewriters.
+    fn is_done(&self) -> bool {
+        // default value indicates we have not completed a cycle
+        let Some(cycle_length) = self.cycle_length else {
+            return false;
+        };
+        self.consecutive_unchanged_count >= cycle_length
+    }
+
+    /// Finishes the iteration by consuming the state and returning a [TreeNode] and
```

Comment:
this finishes the *cycle* right (not the iteration?)

Reply:
> this finishes the _cycle_ right (not the iteration?)
it finishes *all" cycles. as in, it runs all of the remaining iterations and terminates the loop. I will reword to clarify